### ISSUE TITLE
AttributeError: '_VariantDataset' object has no attribute 'output_shapes'

### ISSUE THREAD

**Issue by xsoh** [OpenNMT/OpenNMT-tf#398]
Hi everyone,
I tried the [OpenNMT-tf quickstart](http://opennmt.net/OpenNMT-tf/quickstart.html). I'm running docker tensorflow. When I start training the model I get this error:
```
/usr/local/lib/python3.5/dist-packages/opennmt/config.py:139: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  subconfig = yaml.load(config_file.read())
2019-03-28 23:59:05.257287: W tensorflow/python/util/util.cc:302] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
WARNING: Logging before flag parsing goes to stderr.
I0328 23:59:05.411959 140418515011328 runner.py:98] Using parameters:
data:
  eval_features_file: src-val.txt
  eval_labels_file: tgt-val.txt
  source_words_vocabulary: src-vocab.txt
  target_words_vocabulary: tgt-vocab.txt
  train_features_file: src-train.txt
  train_labels_file: tgt-train.txt
eval:
  batch_size: 32
  eval_delay: 18000
  exporters: last
infer:
  batch_size: 32
  bucket_width: 5
model_dir: run/
params:
  beam_width: 4
  clip_gradients: 5.0
  learning_rate: 0.0002
  length_penalty: 0.6
  optimizer: AdamOptimizer
  param_init: 0.1
score:
  batch_size: 64
train:
  batch_size: 64
  batch_type: examples
  bucket_width: 1
  maximum_features_length: 80
  maximum_labels_length: 80
  sample_buffer_size: -1
  save_summary_steps: 100
  train_steps: 500000

2019-03-28 23:59:05.413641: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-03-28 23:59:05.418977: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2904000000 Hz
2019-03-28 23:59:05.419693: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x4d8f370 executing computations on platform Host. Devices:
2019-03-28 23:59:05.419740: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
I0328 23:59:05.420740 140418515011328 estimator.py:203] Using config: {'_task_id': 0, '_global_id_in_cluster': 0, '_save_checkpoints_steps': None, '_master': '', '_evaluation_master': '', '_experimental_distribute': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7fb592e13cc0>, '_keep_checkpoint_max': 5, '_task_type': 'worker', '_save_checkpoints_secs': 600, '_eval_distribute': None, '_session_config': gpu_options {
}
allow_soft_placement: true
graph_options {
  rewrite_options {
    layout_optimizer: OFF
  }
}
, '_train_distribute': None, '_tf_random_seed': None, '_model_dir': 'run/', '_experimental_max_worker_delay_secs': None, '_num_ps_replicas': 0, '_log_step_count_steps': 100, '_keep_checkpoint_every_n_hours': 10000, '_protocol': None, '_num_worker_replicas': 1, '_device_fn': None, '_save_summary_steps': 100, '_service': None, '_is_chief': True}
I0328 23:59:05.421326 140418515011328 estimator_training.py:186] Not using Distribute Coordinator.
I0328 23:59:05.421565 140418515011328 training.py:612] Running training and evaluation locally (non-distributed).
I0328 23:59:05.422140 140418515011328 training.py:700] Start train and evaluate loop. The evaluate will happen after every checkpoint. Checkpoint frequency is determined based on RunConfig arguments: save_checkpoints_steps None or save_checkpoints_secs 600.
W0328 23:59:05.429433 140418515011328 deprecation.py:323] From /usr/local/lib/python3.5/dist-packages/tensorflow/python/training/training_util.py:238: Variable.initialized_value (from tensorflow.python.ops.variables) is deprecated and will be removed in a future version.
Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.
I0328 23:59:05.481469 140418515011328 data.py:288] Training on 10000 examples
W0328 23:59:05.487812 140418515011328 deprecation.py:506] From /usr/local/lib/python3.5/dist-packages/opennmt/tokenizers/tokenizer.py:277: calling string_split (from tensorflow.python.ops.string_ops) with delimiter is deprecated and will be removed in a future version.
Instructions for updating:
delimiter is deprecated, please use sep instead.
Traceback (most recent call last):
  File "/usr/local/bin/onmt-main", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.5/dist-packages/opennmt/bin/main.py", line 172, in main
    runner.train_and_evaluate(checkpoint_path=args.checkpoint_path)
  File "/usr/local/lib/python3.5/dist-packages/opennmt/runner.py", line 297, in train_and_evaluate
    result = tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_estimator/python/estimator/training.py", line 473, in train_and_evaluate
    return executor.run()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_estimator/python/estimator/training.py", line 613, in run
    return self.run_local()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_estimator/python/estimator/training.py", line 714, in run_local
    saving_listeners=saving_listeners)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 360, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 1140, in _train_model
    return self._train_model_default(input_fn, hooks, saving_listeners)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 1167, in _train_model_default
    input_fn, ModeKeys.TRAIN))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 1004, in _get_features_and_labels_from_input_fn
    self._call_input_fn(input_fn, mode))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 1095, in _call_input_fn
    return input_fn(**kwargs)
  File "/usr/local/lib/python3.5/dist-packages/opennmt/estimator.py", line 124, in _fn
    prefetch_buffer_size=prefetch_buffer_size)
  File "/usr/local/lib/python3.5/dist-packages/opennmt/inputters/inputter.py", line 685, in make_training_dataset
    shard_index=shard_index)
  File "/usr/local/lib/python3.5/dist-packages/opennmt/utils/data.py", line 311, in training_pipeline
    batch_size_multiple=batch_size_multiple))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/ops/dataset_ops.py", line 1746, in apply
    return DatasetV1Adapter(super(DatasetV1, self).apply(transformation_func))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/ops/dataset_ops.py", line 1190, in apply
    dataset = transformation_func(self)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/experimental/ops/grouping.py", line 121, in _apply_fn
    window_size_func)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/experimental/ops/grouping.py", line 377, in __init__
    self._make_reduce_func(reduce_func, input_dataset)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/experimental/ops/grouping.py", line 424, in _make_reduce_func
    input_structure=input_structure)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/ops/dataset_ops.py", line 2400, in __init__
    self._function = wrapper_fn._get_concrete_function_internal()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/eager/function.py", line 1330, in _get_concrete_function_internal
    *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/eager/function.py", line 1324, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/eager/function.py", line 1587, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/eager/function.py", line 1520, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/func_graph.py", line 701, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/ops/dataset_ops.py", line 2394, in wrapper_fn
    ret = _wrapper_helper(*args)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/ops/dataset_ops.py", line 2334, in _wrapper_helper
    ret = func(*nested_args)
  File "/usr/local/lib/python3.5/dist-packages/opennmt/utils/data.py", line 203, in _reduce_func
    return dataset.apply(batch_dataset(batch_size, padded_shapes=padded_shapes))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/ops/dataset_ops.py", line 1190, in apply
    dataset = transformation_func(self)
  File "/usr/local/lib/python3.5/dist-packages/opennmt/utils/data.py", line 143, in <lambda>
    batch_size, padded_shapes=padded_shapes or get_padded_shapes(dataset))
  File "/usr/local/lib/python3.5/dist-packages/opennmt/utils/data.py", line 22, in get_padded_shapes
    return compat.nest.map_structure(lambda shape: shape.as_list(), dataset.output_shapes)
AttributeError: '_VariantDataset' object has no attribute 'output_shapes'
```

**Comment by guillaumekln** [comment]
Hi,

Are you using the nightly Docker image? If yes, please use the latest stable version instead. It is currently `tensorflow/tensorflow:1.13.1-gpu-py3`. Thanks.

**Comment by xsoh** [comment]
worked perfectly, thanks 👍

**Comment by vatai** [comment]
Any idea what could be the problem if I've tried with both `tensorflow/tensorflow:1.13.1-gpu-py3` and `tensorflow/tensorflow:latest-gpu-py3`?

**Comment by guillaumekln** [comment]
Can you provide the steps to reproduce the error using `tensorflow/tensorflow:1.13.1-gpu-py3`?

**Comment by vatai** [comment]
I'm doing all of this on a university cluster, and I don't have root privileges and can't choose software as I wish so `singularity` is the container we have, but it supports docker images (so I think that shouldn't be a problem).

I'm using singularity with the receipt/def file `tf-1.13.1.def` with the following content:
```
Bootstrap: docker
From: tensorflow/tensorflow:1.13.1-gpu-py3

%post
    pip install OpenNMT-tf
```
Then I build the image `tf-1.13.1.sif` from the receipt file:
```
sudo singularity build tf-1.13.1.sif tf-1.13.1.def
```

This is done on my local machine and I transfer the `tf-1.13.1.sif` image file to the cluster. 

On the cluster do the instructed steps from the quickstart:
```
onmt-build-vocab --size 50000 --save_vocab src-vocab.txt src-train.txt
onmt-build-vocab --size 50000 --save_vocab tgt-vocab.txt tgt-train.txt
```
so the files in my directory are
```
batch.sh  src-test.txt  src-train.txt  src-val.txt  src-vocab.txt  tgt-test.txt  tgt-train.txt  tgt-val.txt  tgt-vocab.txt
```
I also add the `data.yml`.

I run the training trough `slurm` by submitting it trough `sbatch` on the following file:
```
#!/bin/sh

#SBATCH -p p
#SBATCH --gres=gpu:2

# LD_LIBRARY_PATH=/usr/local/cuda-10.1/lib64 
srun sh -c "singularity exec --nv tf-1.13.1.sif \
  onmt-main train_and_eval --model_type NMTSmall --auto_config --config data.yml"
```

The only part I'm not sure about is the CUDA version, only 10.1 and 9.1 are available (10.1 is the default).

**Comment by vatai** [comment]
Thank's for the quick response, but it seem's I've figured it out. Singularity (is a great software, but it) mixes badly with `pip install --user`. It saw the `~/.local/lib/python3.5/...` version of tensorflow which was the wrong version!

**Comment by sebastian-nehrdich** [comment]
I get the precise error on my setup now as well. I have Ubuntu 18.04., Cuda 10.0, I built tensorflow from source with CUDA-support (version 1.13.1). I have the latest OpenNMT-tf installed via pip. I am quite puzzled, I already uninstalled the entire CUDA/tensorflow-complex, reinstalled it, I also tried cuda 10.1 (no improvement) and tensorflow 1.12.2 (fails for other reasons). Any suggestions on this?
